walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,169 @@
+  * Then finds the available {@code TimeRange} sandwiched between an existing, ending
+  * {@code TimeRange} and an existing starting {@code TimeRange}.
+  *
+  * @param events Existing events that each occupy some {@code TimeRange}.
+  * @param request A meeting request for some mandatory and optional attendees and a duration.
+  * @return All available {@code TimeRange} that satisfy the meeting request.
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    long requestedDuration = request.getDuration();
+
+    if (Long.compare(requestedDuration, TimeRange.WHOLE_DAY.duration()) > 0) {
+      return Arrays.asList();
+    } else if (events.isEmpty() || attendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<TimeRange> occupiedTimeRanges =
+        getRelevantTimeRangesFromEvents(events, attendees);
+
+    return getAvailabeTimeRanges(occupiedTimeRanges, requestedDuration);
+  }
+
+  /**
+  * Finds the available {@code TimeRange} having at least a length of {@code requestedDuration},
+  * given some already occupied {@code TimeRange}. The resultant list of available
+  * {@code TimeRange} is sorted in ascending chronological order.
+  */
+  private List<TimeRange> getAvailabeTimeRanges(Collection<TimeRange> occupiedTimeRanges,
+      long requestedDuration) {
+    if (occupiedTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<TimeRange> availableTimeRanges = new LinkedList<>();
+    List<TimeRange> startOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                  TimeRange.ORDER_BY_START);
+    List<TimeRange> endOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                TimeRange.ORDER_BY_END);
+
+    // Potentially add the {@code TimeRange} before the first existing {@code TimeRange}.
+    TimeRange firstStartingTimeRange = startOrderedTimeRanges.get(0);
+    checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                          TimeRange.START_OF_DAY, firstStartingTimeRange.start(),
+                                          availableTimeRanges);
+
+    // Find available {@code TimeRange} by looking at the available time between one ending
+    // {@code TimeRange} and the immediately following starting {@code TimeRange}.
+    ListIterator<TimeRange> startOrderedTimeRangesIterator = startOrderedTimeRanges.listIterator();
+    ListIterator<TimeRange> endOrderedTimeRangesIterator = endOrderedTimeRanges.listIterator();
+    TimeRange endingTimeRange = endOrderedTimeRangesIterator.next();
+    TimeRange startingTimeRange = startOrderedTimeRangesIterator.next();
+    while (true) {
```

Comment:
Maybe consider using a
```
do {
  ...
} while (loop break condition);
```

Reply:
Although, there are so many break/continue points that this may not be the cleanest...

Reply:
Maybe consider instead of the "continues" it might be cleaner to use nested while loops? As is, the logic in this function is a bit hard to follow (although not sure if any of my suggestions would make it easier...).  At least it looks like it works, so feel free to ignore my suggestions here unless you feel inspired lol.

Reply:
I modified the code to include nested while loops rather than `continue`s. Hopefully the code has improved readability. I feel that portions of the confusion comes from the bound-checking template code for iterators:
`      if (iterator.hasNext()) {
        item = iterator.next();
      } else {
        break;}`
I might switch to arrays (which allows efficient random access) and manipulate pointers instead.

Reply:
Or, instead of using multiple levels of breaks, you could refactor this out to a helper function and just "return" at the places you do `break findGapsInOccupiedTimeRanges`.

Reply:
Maybe you should switch to using ArrayLists to make accessing the elements clearer?

Reply:
It still seems a bit confusing that you both have variables for the pointers plus the actual timerange object (i.e. that they can be updated independently, potentially). Maybe instead of having the timerange object, just using the orderedTimeRanges[ptr] might be better? +suparnabarua@google.com what do you think?

Reply:
I have that code ready now. I personally think that the code does look clearer

Reply:
yeah that's a good point about using the pointers instead!